torch/nn/parallel/replicate.py

```diff
@@ -2,9 +2,91 @@
 from torch.cuda._utils import _get_device_index
 
 
+def _is_script_module(module):
+    import torch.jit
+    return isinstance(module, torch.jit.ScriptModule)
+
+
+def _init_script_module():
+    import torch.jit
+    return torch.jit.ScriptModule()
+
+
+def _is_jit_enabled():
+    import torch.jit
+    return torch.jit._enabled
+
+
+# Check if we can safely replicate the module.
+# there are three types of module:
+# 1. python modules
+# 2. weak python modules (nn.Module annotated by @weak_module)
+# 3. ScriptModule
+#
+# currently a module cannot be replicated properly if the descendants of
+# any ScriptModule contains python module (type 1 above)
+def _replicatable_module(module, memo=None):
+
+    # module.modules() contains module itself as the first element
+    def descendant_modules(module):
+        gen = module.modules()
+        next(gen)
+        return gen
+
+    if not _is_jit_enabled():
+        return True
+    if memo is None:
+        memo = set()
+
+    memo.add(module)
```

Comment:
this is "visited" flag. It is memorizing that we've visited this module, whether it's legal or not. If any illegal module has been found, it will terminate early and return false. This is why in the second part we can skip any visited modules.